### ISSUE TITLE
Error when trying to initiate debugger: Trying to load Mixfile

### ISSUE THREAD

**Issue by overture8** [JakeBecker/vscode-elixir-ls#18]
I get this error consistently when trying to run the debugger in vscode. Any ideas what would cause this?

```elixir
(Debugger) Initialization failed because an exception was raised:
    ** (Mix.Error) Trying to load Project.Mixfile from "/Users/me/code/project/mix.exs" but another project with the same name was already defined at "/Users/me/code/project/mix.exs"
        (mix) lib/mix.ex:291: Mix.raise/1
        (elixir) src/elixir_module.erl:257: anonymous fn/5 in :elixir_module.expand_callback/6
        (elixir) src/elixir_module.erl:256: :elixir_module.expand_callback/6
        (elixir) src/elixir_module.erl:107: :elixir_module.compile/5
        (stdlib) erl_eval.erl:670: :erl_eval.do_apply/6
        (elixir) src/elixir.erl:239: :elixir.eval_forms/4
        (elixir) src/elixir_compiler.erl:61: :elixir_compiler.eval_forms/3
        (elixir) src/elixir_lexical.erl:17: :elixir_lexical.run/3
```

**Comment by JakeBecker** [comment]
Hmm... My first guess would be that "mix.exs" is included in your launch configuration's `requireFiles` paths, which are intended to pre-load and interpret .exs files you want to debug. That could definitely cause this error, and we should probably exclude mixfiles automatically from it.

Can you check whether your launch config might be including it in `requireFiles`? If that's not it, it'd be helpful if you could post your code somewhere so I can try it.

**Comment by JakeBecker** [comment]
Actually, I don't think that's it. I tried doing that and it didn't reproduce the problem.

Here's my next guess: If VS Code starts the debugger with the cwd set to your project's directory, the Mixfile may be loaded before the task even runs. VS Code doesn't do that on my system (OSX), but it may be a platform dependent thing. What OS are you running?

If that's it, it should be a straightforward fix.

**Comment by JakeBecker** [comment]
Try downloading and installing this updated version of the extension: https://web.tresorit.com/l#2FeWLnm55UscloxhQmlSsQ

If that solves it, I'll do a release. Thanks!

**Comment by overture8** [comment]
Thanks for the quick replies!

I'm actually working on OSX (v10.12.6) as well. I'm not sure if I've missed something in installation. I'm just using the default `launch.json`:

```javascript
{
  // Use IntelliSense to learn about possible attributes.
  // Hover to view descriptions of existing attributes.
  // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
  "version": "0.2.0",
  "configurations": [
    {
      "type": "mix_task",
      "name": "mix (Default task)",
      "request": "launch",
      "projectDir": "${workspaceRoot}"
    },
    {
      "type": "mix_task",
      "name": "mix test",
      "request": "launch",
      "task": "test",
      "taskArgs": [
        "--trace"
      ],
      "projectDir": "${workspaceRoot}",
      "requireFiles": [
        "test/**/test_helper.exs",
        "test/**/*_test.exs"
      ]
    }
  ]
}
```

I installed the updated extension you mentioned and the error has now gone but when I run the `mix (Default task)`, the debugger tool bar appears then disappears without doing anything. I've tried adding a few breakpoints with the same results. Sorry, this is likely just something I'm doing wrong.

**Comment by JakeBecker** [comment]
Can you determine whether the task runs? Maybe add a print statement to the default task so you'll know if it ran or not. Alternatively, try running the test task. The debugger stops after running the task, so if it doesn't hit a breakpoint or print anything, closing silently is pretty much the expected behavior.

**Comment by overture8** [comment]
Here is an example mix project I've set up to test things https://github.com/overture8/testvscode. So, if I open vscode and set a breakpoint here https://github.com/overture8/testvscode/blob/master/lib/testvscode/application.ex#L11 it doesn't seem to pause at this point. Notice the `inspect` above that line - it appears in the Debug Console, so the task is running fine. 

Thanks! 

ps - I'm using Elixir 1.5 and Erlang 20.0.

**Comment by JakeBecker** [comment]
Ah... This is a bit tricky to get right.

Right now, the debugger starts your mix apps before it starts debugging. This is to handle the case where (in Phoenix tests, for example), you're trying to debug .exs files that expect the applications to already be running when they're required. We need to require and interpret any .exs files before launching the task, and since we (sometimes) can't do that without starting the apps first, we start the apps and then handle interpreting files and setting breakpoints. TL;DR: Breakpoints in the application's start process don't work.

I don't know if there's a good one-size-fits-all solution for this. Might need to come up with a way to configure it in `launch.json`. Probably the most compatible way to do it would be to set breakpoints in non-.exs files, launch apps (while awaiting debugger pauses), then require any .exs files and set breakpoints, then run the task.

Can you try it with a breakpoint somewhere in the code called by the task itself? If that works, I'll close this issue and we can make a new one for breakpoints in the application start code not getting hit.

**Comment by JakeBecker** [comment]
I published v0.2.5 with the fix. I'm going to close this issue on the assumption that the original problem with launching the debugger is fixed. If setting breakpoints in the app's initialization code is something you need, feel free to open another issue. Thanks!
